Update Gutenberg packages [MAILPOET-5714]

costasovo opened the pull request:
## Description
This PR updates Gutenberg packages to the versions tagged `wp-6.4` as well as updates of a few packages mentioned in the Dependabot alerts.

## Code review notes
- most of the commits are related to fixing Typescript errors. I'm not sure I addressed all the issues in the best possible way. I'm grateful for any feedback.
- after I fixed all the issues, I faced failing `build_release_zip` jobs. To avoid using a larger instance, I configured build.sh to skip building tests.
- please check also the [PR in the premium plugin](https://github.com/mailpoet/mailpoet-premium/pull/834)

## QA notes

This update affects multiple parts of the plugin as we are using wordpress/components more and more.
The parts that use the Gutenberg packages the most heavily are:

- Form editor
- Automations (all parts)
- Dynamic segments editor
- New email editor
- Coupons block in the current email editor

Please test these parts thoroughly.

## Linked PRs

https://github.com/mailpoet/mailpoet-premium/pull/834

## Linked tickets

[MAILPOET-5714]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5714]: https://mailpoet.atlassian.net/browse/MAILPOET-5714?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

costasovo committed:
Fix types issues in font selection format

The new version of @wordpress/rich-text comes with build types,
but the RichTextFormat is missing attribute values.
[MAILPOET-5714]

costasovo committed:
Fix 'useSelect' issue causing multiple renders of basic settings

This commit fixes a warning:
"The 'useSelect' hook returns different values when called with the same state and parameters. This can lead to unnecessary rerenders."
The warning was coming from the BasicSettingsPanel component and was caused
by getAllAvailablePages selector, which returned a new array every time.

[MAILPOET-5714]

costasovo committed:
Refactor multiple useSelect calls into one in BasicSettingsPanel

[MAILPOET-5714]

costasovo committed:
Update deprecated Button props

[MAILPOET-5714]

costasovo committed:
Remove unused WordPress module declarations

These were needed for the types for stores used in the email editor
when it was based on the post editor. We switched to the custom editor
so we can remove those.
[MAILPOET-5714]

costasovo committed:
Fix types for Notices

There is a mismatch between the WPNotice type and the WordPress/notice store
and Notice defined in the components package.

[MAILPOET-5714]

costasovo committed:
Remove custom type definitions that were added in packages

[MAILPOET-5714]

costasovo committed:
Use proper types for Color and FontSize

[MAILPOET-5714]

costasovo committed:
Fix types for SelectControl Options

[MAILPOET-5714]

costasovo committed:
Fix remaining Typescript warnings in form-editor

[MAILPOET-5714]

costasovo committed:
Remove unnecessary Popover.Slot

It is not longer needed see: https://github.com/WordPress/gutenberg/pull/53889
It was causing TS errors
[MAILPOET-5714]

costasovo committed:
Fix SearchControl error in Segments Template Selection

The search functionality hasn't been implemented yet. The task is currently in backlog.
The control was added in an initial task but is hidden.
I decided to keep it so that I don't remove the translation and just add a dummy
onChange callback to prevent TS error.
[MAILPOET-5714]

costasovo committed:
Update Button component props type in MailPoet automation integration

[MAILPOET-5714]

costasovo committed:
Fix TS errors in assets/js/src/common

[MAILPOET-5714]

costasovo committed:
Fix ToolbarItem component typescript errors

[MAILPOET-5714]

costasovo committed:
Make useDispatch hook type in settings store compatible with wp/data

[MAILPOET-5714]

costasovo committed:
Workaround TS error about missing variant prop on MenuItem component

[MAILOPET-5714]

costasovo committed:
Remove passing setIsBusy to dropdown control in OptionButton

Overriding onClick: (setIsBusy?: Dispatch<SetStateAction<boolean>>) => void;
was causing TS errors when passing controls with the type ControlWithSetIsBusy to DropdownMenu.

The OptionButton that uses isBusy logic is used only in the Hero section at automation
Homepage. The setIsBusy function that was passed to control's onClick was not used.

In all other cases where we were passing controls to DrowdownMenu, we didn't set any isBusy state.
So I think it is safe to remove it.
[MAILPOET-5714]

costasovo committed:
Remove unnecessary fix for missing Popover props

The new type has the props.
[MAILPOET-5714]

costasovo committed:
Fix error JSX element type ConfirmDialog does not have any construct or call signatures

[MAILPOET-5714]

costasovo committed:
Fix buttons prop types issues in automation

[MAILPOPET-5714]

costasovo committed:
Update the type of the button with extended props used in automation

[MAILPOET-5714]

costasovo committed:
Fix value in Dropdown focusOnMount prop

The currently supported values are 'firstElement' | boolean
Based on the JS doc and my testing, we need to use true
to hide the popup on click elsewhere or by ESC.
[MAILPOET-5714]

costasovo committed:
Fix the rest of TS errors in automations codebase

[MAILPOET-5714]

costasovo committed:
Fix document actions appearance in automations editor

The dropdown component used for setting the automation name is based
on the document actions component from edit-site package. After the update,
the appearance changed significantly, and it was broken.
This commit adds CSS to restore the original look as before the update.
[MAILPOET-5714]

costasovo committed:
Fix incorrect indentation of the admin panel in automations editor

[MAILPOET-5714]

costasovo committed:
Fix TS issues with props in automation TemplateDetail component

[MAILPOET-5714]

costasovo committed:
Fix eslint issues

The issues were related to a change in Gutenberg where ActionCreators are not promisified.
See https://github.com/WordPress/gutenberg/pull/52530
[MAILPOET-5714]

costasovo committed:
Update WordPress packages to tag wp-6.4

This commit sets all wordpress/* packages we list in package.json
to use tag wp-6.4, which should install the version used in WordPress 6.4
[MAILPOET-5714]

costasovo committed:
Remove @types/wordpress__ for packages that have build-types

[MAILPOET-5714]

costasovo committed:
Update @types/wordpress_xxx packages

[MAILPOET-5714]

costasovo committed:
Remove unnecessary wrapper for toggle sidebar panel in form editor

[MAILPOET-5714]

costasovo committed:
Opt-in to no margin for FontSizePicker

The bottom margin will be deprecated. With __nextHasNoMarginBottom
switched to the future styles without the margin.
[MAILPOET-5714]

costasovo committed:
Update wordpress/* dev dependencies also to wp-6.4

[MAILPOET-5714]

costasovo committed:
Update browserify-sign dependency

[MAILPOET-5714]

costasovo committed:
Update acceptance tests

[MAILPOET-5714]

costasovo committed:
Remove unnecessary types overrides for Wordpress/components

I believe these are no longer in use because wordpress/components
come with built types.
[MAILPOET-5714]

costasovo committed:
Disable TsCheckerPlugin for email editor production build

[MAILPOET-5714]

costasovo committed:
Add support for skipping build of JS tests and skip when building zip

This commit addresses issue with insufficient resources on Circle CI
to avoid using larger executor instances.
[MAILPOET-5714]

costasovo committed:
Fix error for Tooltip text using ReactNode

[MAILPOET-5714]

costasovo committed:
Fix support of WP admin colors schemes for all plugin pages

When checking pages after updating dependencies, I noticed that some buttons have strange colors.
For example, buttons on the "What would you like to create?" page for new newsletters.

I found that this was related to a change in wordpress/components styles.
The original colors were identical to the default WordPress admin color scheme,
so the buttons were ok when an admin used the default scheme, and in case they use a different scheme
the buttons still look like default (blue).
But after the update, all components on MailPoet pages that don't load WP color admin variables
had a strange purple color.

This commit fixes it by including CSS for admin color schemes into mailpoet-plugin.css which is loaded on every MailPoet admin page.

Note the mixin wordpress-admin-schemes() is included in many build styles for other wordpress packages (e.g. wordpress/block-editor)
so admin color schemes worked correctly on some MailPoet pages even before this change.
[MAILPOET-5714]

JanJakes commented on mailpoet/assets/js/src/automation/editor/components/header/inserter-toggle.tsx:
```diff
@@ -1,9 +1,17 @@
+// Build type for ToolbarItem contains only "as" and "children" properties but it takes all props from
+// component passed to "as" property (in this case Button). So as fix for TS errors we need to pass all props from Button to ToolbarItem.
+// We should be able to remove this fix when ToolbarItem will be fixed in Gutenberg.
+const ToolbarItem = WpToolbarItem as React.ForwardRefExoticComponent<
```
I only wonder what exactly is `ForwardRefExoticComponent` and where it comes from. Did you use it because it was in the original type? Or is it the only variant that works?

costasovo replied:
@JanJakes I had quite a hard time with some of those Typescript errors. I'm not sure I remember exactly, but in this case, I think yes I used it because it was in the original build types. From that, I saw that I could add another component, and it would accept props. In this place, I did it only for Button.

`/node_modules/@wordpress/components/build-types/toolbar/toolbar-item/index.d.ts`

```ts
/// <reference types="react" />
declare const _default: import("react").ForwardRefExoticComponent<Omit<import("react").HTMLAttributes<any>, "children"> & {
    as?: import("react").ElementType | undefined;
    children?: import("react").ReactNode | ((props: import("react").HTMLAttributes<any> & import("react").RefAttributes<any>) => import("react").ReactElement<any, string | import("react").JSXElementConstructor<any>> | null);
} & import("react").RefAttributes<any>>;
export default _default;
//# sourceMappingURL=index.d.ts.map
```

JanJakes commented on mailpoet/assets/js/src/automation/editor/components/actions/trash-button.tsx:
```diff
@@ -1,12 +1,16 @@
 
+// With __experimentalConfirmDialog's type from build-types Typescript complains:
+// JSX element type __experimentalConfirmDialog does not have any construct or call signatures
+// Wrapping the type to React.FC fixes the issue
```
I wonder if this was actually useful in some form: https://github.com/mailpoet/mailpoet/pull/5351/commits/d66769e80ed93ac5570c223961682e225e5fa8d9#diff-f121ab46250da530ff0e99594c24c398d2c4527aeab72d5b9595b8ed7e3fc28bL39

costasovo replied:
I think that after the update, these were ignored because TS took the definition from build types.

JanJakes commented on mailpoet/assets/css/src/components-automation-editor/_dropdown.scss:
```diff
@@ -1,3 +1,39 @@
+.mailpoet-automation-editor-edit-site-document-actions {
```
Here I'm wondering:

1. Should we use the new look to align with Gutenberg, or it doesn't work in this case?
2. Should we create our custom component for this?

JanJakes replied:
I think previously, it looked a bit different.

Now:
<img width="312" alt="Screenshot 2024-01-17 at 12 34 07" src="https://github.com/mailpoet/mailpoet/assets/141436/ffff90b3-0cc7-4ac9-933d-05ff0991fee8">

Before:
<img width="310" alt="Screenshot 2024-01-17 at 12 35 54" src="https://github.com/mailpoet/mailpoet/assets/141436/e72d711b-552d-45d0-a8f2-2f021d9272f4">

costasovo replied:
@JanJakes, I'll update CSS  so that the result is closer to the original.

> Here I'm wondering:
> 
> Should we use the new look to align with Gutenberg, or it doesn't work in this case?
> Should we create our custom component for this?

Good questions. 

The new Gutenberg design is something completely different. Click on the component triggers the command palette.

<img width="537" alt="Screenshot 2024-01-19 at 15 26 08" src="https://github.com/mailpoet/mailpoet/assets/1082140/38031022-0ec4-4e25-9a28-496b7a50049f">

As I see it, what we currently have is not aligned with Gutenberg (design or purpose), so we should create a custom component. Or even discuss with the design team if we still want to use that area (center of the header) for this component as it is now not aligned with Gutenberg.

Do you want me to create the component within this PR? I would suggest going with the CSS fix for now so that we no longer block this PR and I would prepare a ticket about refactoring the component. Would you be ok with that?

JanJakes commented on mailpoet/webpack.config.js:
```diff
@@ -452,7 +452,9 @@ const emailEditorCustom = Object.assign({}, wpScriptConfig, {
-  plugins: [...wpScriptConfig.plugins, ...[new ForkTsCheckerWebpackPlugin()]],
+  plugins: PRODUCTION_ENV
+    ? wpScriptConfig.plugins
+    : [...wpScriptConfig.plugins, ...[new ForkTsCheckerWebpackPlugin()]],
```
Very unimportant, but this is the same: `[...wpScriptConfig.plugins, new ForkTsCheckerWebpackPlugin()]`

Or maybe it was meant to be like this?

```ts
plugins: [...wpScriptConfig.plugins, ...(PRODUCTION_ENV ? [new ForkTsCheckerWebpackPlugin()] : [])],
```

Then the spread would make sense...

JanJakes commented on mailpoet/build.sh:
```diff
@@ -15,7 +15,7 @@ mkdir $plugin_name
 rm -rf node_modules
 pnpm install --frozen-lockfile --prefer-offline
-./do compile:all --env production
+./do compile:all --env production --tests skip
```
Should we maybe use `--skip-tests` or `--skip-tests-build` to make it more unified with `--skip-deps` for tests?

JanJakes replied:
I also wonder if it could work based on `env`. Do we ever build tests with `production` env?

costasovo replied:
I like the idea of unifying the naming! 

As for making it dependent on production I saw that [on CI for the JS tests we use the production env](https://github.com/mailpoet/mailpoet/blob/db9813a3aac5e1ebfd4fd3d692ab54f0ed95bb5f/.circleci/config.yml#L157). When I searched node_modules for `process.env.NODE_ENV` I saw that it is used in many cases to check if the value is `production` so I assume the production environment has additional effects. So I didn't want to change that we run the tests build with the production settings.

JanJakes commented on mailpoet/assets/js/src/types/index.ts:
```diff
@@ -1,9 +1,7 @@
-import { ComponentPropsWithoutRef } from 'react';
```
Related to this commit, https://github.com/mailpoet/mailpoet/pull/5351/commits/6d842ec0f0342bde89d01de78bacb345b7bdfd64 doesn't remove `@types/wordpress__components`. Maybe we discussed this, but are now both needed?

costasovo replied:
@JanJakes Ha. Thanks for mentioning the `@types/wordpress__components` I see I haven't removed the package. I skipped removing it because it completely broke the premium build. Somehow, the imports from `@wordpress/components `that are exposed from the free plugin don't work if I remove the package. I had a lot of issues to fix and completely forgot about this one. I'll try to look into it. 

Also, I don't understand what you are reffering to in "Maybe we discussed this, but are now both needed?" What is the second thing?

JanJakes commented on mailpoet/RoboFile.php:
```diff
@@ -202,7 +202,7 @@ public function compileJs($opts = ['env' => null]) {
       sprintf('./node_modules/.bin/cross-env NODE_ENV="%s"', $opts['env']) :
       null;
-    return $this->_exec($env . ' ./node_modules/webpack/bin/webpack.js');
+    return $this->_exec($env . ' ./node_modules/webpack/bin/webpack.js --env BUILD_TESTS=' . ($opts['tests'] === 'build' ? 'build' : 'skip'));
```
On my machine, `./do compile:js` now fails (`./do compile:all` works):

<img width="430" alt="Screenshot 2024-01-17 at 12 37 45" src="https://github.com/mailpoet/mailpoet/assets/141436/8f3db502-5a53-492f-b842-e8716f58923c">

JanJakes commented on mailpoet/assets/css/src/mailpoet-plugin.scss:
```diff
@@ -18,6 +18,11 @@
+// Needed for the WordPress Admin color scheme support for @wordpress/components.
+@import '../../../node_modules/@wordpress/base-styles/mixins';
+@include wordpress-admin-schemes();
+
```
I'm still seeing strange colors on the homepage:

<img width="475" alt="Screenshot 2024-01-17 at 12 39 04" src="https://github.com/mailpoet/mailpoet/assets/141436/41142ccd-1b38-48df-a7be-407520eacc5a">

I installed all deps and compiled all assets, caching in browser disabled (dev mode).

JanJakes commented on mailpoet/assets/js/src/automation/integrations/mailpoet/components/button.tsx:
```diff
@@ -1,13 +1,27 @@
+      Omit<ButtonAsAnchorProps, keyof ExtendedProps> & ExtendedProps,
+      'a',
+      false
+    >;
```
The issue here is that the Button props are a union of multiple types. We need to apply the modifications (omit + add new props) to all the union members. That's more universally doable using [distributive conditional types](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html#distributive-conditional-types):

```ts
type UpdateUnionProps<T, U> = T extends unknown ? Omit<T, keyof U> & U : never;

type Props = UpdateUnionProps<
  React.ComponentProps<typeof WpButton>,
  ExtendedProps
>;
```

costasovo replied:
I used your definition. I need to take a look at it again with a fresh head to fully understand how that works.

JanJakes replied:
@costasovo It basically means "for each member of union T, do `Omit<T, keyof U>`".

JanJakes reviewed (commented):
Great job! I have only a few comments.

costasovo committed:
Fix padding in automation editor name edit popup

[MAILPOET-5714]

costasovo committed:
Rename flag for skipping tests in webpack build

[MAILPOET-5714]

costasovo committed:
Remove unnecessary array spread in webpack config for email editor

[MAILPOET-5714]

costasovo committed:
Use conditional types for fixing button type

[MAILPOET-5714]

costasovo committed:
Remove @types/wordpress__components package

[MAILPOET-5714]

costasovo committed:
Fix lint issue after rebase

[MAILPOET-5714]

costasovo commented:
@JanJakes I believe I addressed all your comments. I rebased the PR and added 7 new commits (beginning with https://github.com/mailpoet/mailpoet/pull/5351/commits/acab2ab80d04b06c72a2be0cfde210dbc6f96fb0). 

After I removed the `@types/wordpress__components` I needed to do additional changes in the [related premium-plugin PR](https://github.com/mailpoet/mailpoet-premium/pull/834)

Aschepikov commented:
Hi @costasovo I found some small UI issues, can you take a look pls
1 In the new email editor, there is no slider for the padding and no background for the button, the current prod is on the right side
![image](https://github.com/mailpoet/mailpoet/assets/113185437/58e3378a-e373-4112-80ca-425c6ffa0dda)

2 In the segments, some of the conditions are too wide and overlap with the remove icon, on the prod everything is ok
![image](https://github.com/mailpoet/mailpoet/assets/113185437/5ebe546a-d647-4d15-bac6-26e61b36de40)

3 In the automation analytics, there is some overlap in date range setting
![image](https://github.com/mailpoet/mailpoet/assets/113185437/8d8c2f59-4846-44bd-b55e-ac425b5d89c5)

4 Any click on the date or the month scroll closes the date popover (automation analytics)
![image](https://github.com/mailpoet/mailpoet/assets/113185437/911b0da6-f1a5-4368-a509-b81c34e41d9d)

5 The gradient isn't saving and after reloading the page there is no gradient (form editor)
![image](https://github.com/mailpoet/mailpoet/assets/113185437/b95043fa-5b8e-43d7-9eb7-a95a41deb845)
![image](https://github.com/mailpoet/mailpoet/assets/113185437/f390e2fd-80b7-43a1-86e0-438cfff1f967)
![image](https://github.com/mailpoet/mailpoet/assets/113185437/262d815e-b630-4050-b216-e43f212e0720)

costasovo committed:
Add spacing sizes presets

This commit addresses an issue that padding control in the email
an editor is rendered without a slider.
This change makes the padding control render the same as in TT4 theme
[MAILPOET-5714]

costasovo committed:
Fix apperance of date filter in automation analytics

All popups are not appended at the end of body so we needed to
adjust CSS selectors that were used to modify the component.
[MAILPOET-5714]

costasovo committed:
Add support for gradient for heading and paragraph in form editor

[MAILPOET-5714]

costasovo committed:
Add gradient support for heading and paragraph to form renderer

[MAILPOET-5714]

costasovo committed:
Add support for padding for headings and paragraphs in the form editor

[MAILPOET-5714]

costasovo committed:
Add support for heading and paragraph padding to form renderer

[MAILPOET-5714]

costasovo commented:
Thank you, @Aschepikov, for the thorough review! I believe I addressed all the issues.

> 1 In the new email editor, there is no slider for the padding and no background for the button, the current prod is on the right side

I fixed that by configuring the editor to provide steps for setting padding. This will make it more comfortable for users and they still have an option to insert exact units.

> 2 In the segments, some of the conditions are too wide and overlap with the remove icon, on the prod everything is ok

Fixed by adding a CSS for the trash bin icon. The CSS was added to the premium plugin because we allow multiple conditions only with premium. So please test with the build from the premium plugin PR https://github.com/mailpoet/mailpoet-premium/pull/834

> 3,4,5 

Should be fixed.

@JanJakes I rebased the PR please check the last 6 commits.

Aschepikov commented:
Hi @costasovo  I am rechecking everything and I see in the automation analytics that there is a problem with step icons on the subscriber tab
![image](https://github.com/mailpoet/mailpoet/assets/113185437/1ec7912b-927e-420d-a084-0a5b93ff1407)
and on the automation flow too
![image](https://github.com/mailpoet/mailpoet/assets/113185437/08f70ca0-d3b4-48eb-a896-8c86dc916086)
Also, I have a problem with the edit automation button. I have this on all browsers but there was a case when it wasn't reproducible on the dev PC so I am not sure about it
![image](https://github.com/mailpoet/mailpoet/assets/113185437/c3e141be-d2d6-4629-9c0d-caad1028b255)

costasovo committed:
Fix icons and spacing issues on the automations stats page

With the previous fix I accidentally applied box-sizing: border-box;
to everything on the page.
This commit fixes it so that the rule is applied only in places
where it was applied before the upgrade.
[MAILPOET-5714]

costasovo commented:
@Aschepikov Nice catch! It was caused by my fix of filters in Automation analytics 🤦 
@JanJakes I added one new commit with a fix. Could you please take a look? 🙏

Aschepikov commented:
Ho @costasovo  Great, everything looks good but I see again the 4th issue "Any click on the date or the month scroll closes the date popover (automation analytics)". So with the month scroll buttons, it is impossible to change the month as it closes after every click. The same with the date range but the changes are saving. On the prod, I don't have it. Sorry, I hope it is the last call 
![image](https://github.com/mailpoet/mailpoet/assets/113185437/c3471c9a-51f8-4c1a-a34f-4c4fe891e1e3)

costasovo committed:
Fix deprecation warnings for popovers

[MAILPOET-5714]

costasovo committed:
Dont render an empty tooltip for failed stats in mailpoet:send-email step

This fixes a console warning: "Tooltip should be called with only a single child element."
[MAILPOET-5714]

costasovo committed:
Fix issue with month switching in date range picker

The issue is related to the underlying react-dates library used
in woocommerce/components in calendar/date-range.js

See https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/components/src/calendar/date-range.js#L45-L104

I was not able to fully understand why it started to happen after the update
but I have a suspicion that it may be related to the change that causes rendering
popups at the end of the body.

The Gutenberg library already moved away from react-dates and I assume
that Woo Components will do that as well.

This commit adds a patch that adds e.stopPropagation() to the place
where the original workaround is trying to prevent losing focus.

[MAILPOET-5714]

costasovo commented:
@Aschepikov My bad! I missed the issue and thought all in the picker was related to the broken visual appearance.  It should be fixed now.

@JanJakes One more review, please 🙏  I added three more commits. I noticed a couple of warnings in the console on the automation stats page. There are still few warnings left. These were [fixed in woocommerce/components](https://github.com/woocommerce/woocommerce/pull/41845) but we need to update woocommerce/components.

Aschepikov commented:
HI @costasovo on the subscriber's tab in the automation analytics the icons are non-centrical in the circle
![image](https://github.com/mailpoet/mailpoet/assets/113185437/c7fa3dbe-cedb-484d-be82-795ed12babe9)

costasovo commented:
> HI @costasovo on the subscriber's tab in the automation analytics the icons are non-centrical in the circle

@Aschepikov Is this happening in a specific browser? I tested it in Chrome, Firefox, and Safari. I also tried using the zip from Circle CI, but in all cases, the icon was centered correctly.

Aschepikov commented:
I have it in explorer and chrome but it can be connected with the site after a lot of installs, it should be fixed with a release build as you don't have this, thank you

Aschepikov reviewed (approved):
GJ!
